### DISCUSSION TITLE
Create a Github-hosted PPA for Jamulus

### DISCUSSION THREAD

**Discussion by pljones** [jamulussoftware/jamulus#2122] [Ideas]
**Background**

x64 Linux users with distributions based on Debian currently need to download and install the `.deb` file manually from each Release of Jamulus if they want to keep up to date.  This is because most of the Debian-based distributions are falling behind the Release version, as they're only maintained on a best-endevours basis.

The usual mechanism used in cases such as this is a Personal Package Archive (PPA) referencing the software.  This allows the user to keep the software update using the built in tools for their distribution.

**Proposal**

It is possibly to maintain a PPA directly on Github.  The process is explained by Assaf Morami [here](https://assafmo.github.io/2019/05/02/ppa-repo-hosted-on-github.html).  It assumes that a single Github user account would maintain the PPA but I am proposing that the PPA for Jamulus be "owned" by the Jamulussoftware organisational account.  The release workflow would take place as is, ending up with a new tagged release.  This would then need to trigger (ideally automatically) a workflow in the PPA repository to pick up the release `.deb` file(s) and add them to the PPA.

This would require a GPG key to be created on behalf of the account by a "trusted party" and then secured such that only the Jamulussoftware account and any Github accounts trusted to manage that account had access to the private (signing) key.  Any existing Jamulussoftware account Admin could create the GPG keypair for Jamulussoftware.  Each Admin would also need provide their GPG _public_ key to the key creator so that they could received an encrypted copy of the keypair that only they could import.

(There's also fluff around key-signing keys and key revokation that would need thinking about.)

Github has provision for such secrets.  We could use a repository-level secret that is recognised only as the signer for the PPA repo.
https://docs.github.com/en/actions/security-guides/encrypted-secrets

There is a market place action to bring the secrets into the workflow.
https://github.com/marketplace/actions/import-gpg

**Comment by ann0see** [comment]
Great idea! Actually I also thought about this, but failed due to the lack of knowledge in the automated setup.
A few thoughts:
> This would then need to trigger (ideally automatically) a workflow in the PPA repository to pick up the release .deb file(s) and add them to the PPA.

I'm unsure if it's possible to link two repositories with GitHub actions. I think you could probably push from the main repo to the PPA repo by using a bash script. So the workflow would look like that:

1. Push tag to GH (maindev)
2. Autobuild deb
3. git push deb to PPA repo (the "automated" step)
4. On the PPA repo a GH actions script would run (since it received a push) and rebuild all the files necessary to build update the PPA
5. The new release would be accessible via the PPA

Concerning the signing, someone who is deeper into this topic should have a look at it (@hoffie)?

**Reply by ann0see** [reply]
Did a quick check on adding a new repository (debianppa). However you'd need to switch on this checkbox: https://github.com/organizations/jamulussoftware/settings/member_privileges#create_public_pages_enabled

**Reply by pljones** [reply]
> 4.  On the PPA repo a GH actions script would run (since it received a push) and rebuild all the files necessary to build update the PPA

Yes, something like that is what I was thinking -- just giving it a kick some how.  The build process for jamulussoftware/jamulus would need push permission on jamulussoftware/jamulusppa (or whatever).

But why would the Create Public Pages be needed?

**Reply by ann0see** [reply]
To get a the repo served as GH page? Didn’t read the link yet though

**Reply by ann0see** [reply]
@pljones @softins @gilgongo how can we move this forward?
Create a GPG key?

**Reply by gilgongo** [reply]
As I understand it, the admins need to first make sure they trust each other's public keys. I'm not entirely sure what that would involve. Some of us have met over Zoom before, so I'd recognise them, while others I haven't. Does that imply showing each other passports or answering questions only they would know while reading out GPG fingerprints or something?

But if we can all be satisfied that the keys we have received are good, then one of us creates a key pair with a passphrase, encrypts that passphrase to the other admins, and sends the private key to them by other means. 

At that point any admin can sign a release as per Assaf's guide, is that right?

**Reply by pljones** [reply]
Given we've all got Github accounts, we can put the public key to our keypair on Github (if we're using 2FA, like we're meant to).

Then key exchange can happen and verification is by confirming the signature using the public key on the associated Github account.  We _must_ already **trust** Github to identify of each of us...

**Reply by gilgongo** [reply]
OK I've got my key on GH as part of 2FA, so what next?

**Reply by pljones** [reply]
https://blog-eleven--labs-com.translate.goog/fr/openpgp-clef-participe-a-une-fete-de-la-signature-des-clefs/?_x_tr_sl=fr&_x_tr_tl=en&_x_tr_hl=en-GB

We want a `jamulussoftware` key pair.

We want each of the main developers to sign it.

But we also want each of the main developers to sign each other's keys.

This only really matters if the `jamulussoftware` public key gets pushed to a server, as we'd want our public keys pushed, too, to authenticate it.  That "ring of trust" is how OpenPGP works.

Once the `jamulussoftware` key pair exists, it's used to sign the PPA content.

In terms of ID, I'm not sure how much "in person" contact helps.  For example, I've got a LinkedIn account with my details on linked to my personal accounts, including Github (now...). https://www.linkedin.com/in/peterljones/  -- I guess I could supply an image of my work pass as a "handshake" on that?  (Mmm, it might even be the same photo...)

The idea behind having the 2FA Github and the pubkey there, is that gives some trust already -- which maybe skips that need for "in person" contact -- or even any further verification.  Like I say, if you've access to your Github account to put a pubkey on it, you're trusted enough to push releases out the door as it is, so what further verification is needed?

**Reply by gilgongo** [reply]
Apologies, what I meant when I said "What next?" was what _exactly_ is next? Shall I create a repo under the org? Are GH "secrets" the same thing as key pairs? If so, what do we/I/you need to do now? Etc. :-) In short: how can I help?

**Reply by pljones** [reply]
All I've got are what's in the items I've linked.  Unfortunately, there's no organisation example that I've found, so one thing to work on would be to see if something like that exists anywhere and I just missed it.

**Reply by gilgongo** [reply]
Oh OK. Does the PPA for Jamulus definitely need to be "owned" by the Jamulussoftware organisational account, or can we just have a signing key for the PPA repo? When I tried setting up a GH action to put new issues and PRs from more than one repo automatically into the project we have, organisation permissions seem extremely hard to work out (in fact I gave up, and so did the various people who tried to help from the market place scripts I was using).

**Reply by hoffie** [reply]
> I say, if you've access to your Github account to put a pubkey on it, you're trusted enough to push releases out the door as it is, so what further verification is needed?

I agree. The purpose of the GPG signing in this case is proof that the packages come from us. While signing the team key with our individual keys would certainly be nice, I don't think it should be a requirement. Providing a simple repo and the relevant pubkey via https from one of our established places (= Website, or Github directly) should be sufficient to provide the same level of trust and security which we have for ordinary release downloads.
What we do have to think through though is GPG key lifetime and possibly key rotation procedures.

I think repo creation is perfectly doable. There are lots of examples, scripts and actions, although lots of it seems dated and possibly unmaintained.

As far as I can see, there are currently two approaches for hosting such a repo:
1. Github Pages
1. Github Releases

Pages sounds more natural to me (possibly because it most closely resembles "just a webserver", which would be needed when doing things manually).

Regarding triggers, I see the following options:
1. Build the repo as part of the jamulus repo and upload it as artifact there or publish it as Github page from there. This way, `GITHUB_TOKEN` should work.
1. Use a separate repo for the workflow and the contents. `GITHUB_TOKEN` will not work there as it is bound to the source repo.
  a. We can either use a machine account (a regular Github account) with appropriate permissions and a matching Personal Access Token,
  b. We can create a synthetic Jamulus Github App (as pointed out by @dtinth, thanks!): https://github.com/peter-murray/workflow-application-token-action or
  c. We can push the button manually (`workflow_dispatch`). The latter may sound non-optimal wrt automation. However, lots of roll-out strategies include rolling out to a small subset of users first. So maybe it would not be a bad idea at all to start publishing via repos only a few days after the official release in order to give the final release some real world testing by non-repo users?

I may have some time to do further tests regarding this topic and will report back here.

**Reply by dtinth** [reply]
I think it's also possible to start from (c), then add in either (a) or (b) later.

For example:

1. Create a `ppa` repo to store the `.deb` files.
2. Create a manually-triggered GitHub Actions workflow that generates `Packages`, `Packages.gz`, `Release`, `Release.gpg`, `InRelease` files and push it to `gh-pages` branch, thus publishing the ppa. Everything happens within the repo, so `$GITHUB_TOKEN` works out-of-the-box.
3. To automatically release, the `jamulus` repo can use [benc-uk/workflow-dispatch](https://github.com/benc-uk/workflow-dispatch) action can be used to trigger a workflow on the `ppa` repo.

To ensure integrity of dependencies, can do one of two things:

- Depend on a specific hash, i.e. `uses: benc-uk/workflow-dispatch@4c044c1613fabbe5250deadc65452d54c4ad4fc7 # v1.1.0`
- Fork the action into Jamulus’ org and use that instead, i.e. `uses: jamulussoftware/workflow-dispatch@v1`

**Comment by ann0see** [comment]
Just found reprepro: https://salsa.debian.org/brlink/reprepro/-/blob/debian/docs/short-howto the package seems to be quite interesting (and not too difficult to implement). I mean at first, we could provide a beta, un-signed PPA.

**Reply by pljones** [reply]
I tried to read the "short-howto" but it left a lot of unanswered questions without direction to where to get the answers...  I guess if you already know what it's talking about, it might make more sense.

**Reply by ann0see** [reply]
Yes, it seems a bit confusing, but as far as I read we‘d need to create a configuration file, import a GPG key (using the action you mentioned) and that’s it. 

This German wiki site is maybe also interesting https://wiki.ubuntuusers.de/reprepro/

**Comment by ann0see** [comment]
Hoffie did some work on https://github.com/hoffie/jamulus-repos and I've imported + changed his repo and pushed it to https://github.com/jamulussoftware/repos/ 

There's also a slightly different approach I'm thinking of in https://github.com/jamulussoftware/repos/issues/2 

In short: Change the autobuild process to upload the metadata files directly to the latest release after they are finished and rely on GitHub redirects. The Repo URL would be e.g. https://github.com/jamulussoftware/jamulus/releases/latest/download/

**Reply by ann0see** [reply]
Ok. My approach is now here: https://github.com/jamulussoftware/jamulus/pull/3013 (including work by hoffie)
